app/src/org/commcare/android/resource/installers/OfflineUserRestoreAndroidInstaller.java

```diff
@@ -0,0 +1,74 @@
+package org.commcare.android.resource.installers;
+
+import org.commcare.CommCareApplication;
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.models.database.user.DatabaseUserOpenHelper;
+import org.commcare.resources.model.Resource;
+import org.commcare.resources.model.ResourceInitializationException;
+import org.commcare.resources.model.UnresolvedResourceException;
+import org.commcare.suite.model.OfflineUserRestore;
+import org.commcare.utils.AndroidCommCarePlatform;
+import org.javarosa.core.reference.InvalidReferenceException;
+import org.javarosa.core.reference.Reference;
+import org.javarosa.core.services.storage.EntityFilter;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.core.util.externalizable.PrototypeFactory;
+import org.javarosa.xml.util.InvalidStructureException;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * @author Phillip Mates (pmates@dimagi.com)
+ * @author Aliza Stone (astone@dimagi.com)
+ */
+public class OfflineUserRestoreAndroidInstaller extends FileSystemInstaller {
+
+    @SuppressWarnings("unused")
+    public OfflineUserRestoreAndroidInstaller() {
+        // for externalization
+    }
+
+    public OfflineUserRestoreAndroidInstaller(String localDestination, String upgradeDestination) {
+        super(localDestination, upgradeDestination);
+    }
+
+    @Override
+    public boolean initialize(AndroidCommCarePlatform instance, boolean isUpgrade) throws ResourceInitializationException {
```

Comment:
So this is a tricky point, but I don't know that the semantics of this method work quite right for this case.

the isUpgrade flag here lets the app know to process the initialization in-place because there was an upgrade, but it is 

A) Not guaranteed to fire (IE: an upgrade can "complete" in the engine, and then an error can restart the process, and the app will initialize without firing the isUpgrade version of this method)
B) Only happens after an update is finalized and has become the current running mode

B is important because it means that even if this method was required to fire before the app could proceed, if there were an error in the demo data, the app could never get back on track, because an error in the restore data would prevent the update from continuing.

Any work that has to be guaranteed to happen has to occur in the customInstall or root install methods. or their lifecycle needs to somehow reflect its own guarantees that something was processed.